### DISCUSSION TITLE
Automatically Generating Part Numbers

### DISCUSSION THREAD

**Discussion by harryg-oss** [inventree/InvenTree#6151] [Ideas]
Hi,

I am looking to see if there exists a solution to automatically generate part numbers, perhaps from a pattern defined in the subclasses.
Ideally I would like a pattern input when creating/editing the subclass (which would need a new column in the database by the looks of it), and a button next to the input box when creating/editing a part to automatically generate the part number.

I see there is some pattern input for Global Settings -> Stock -> Batch Code Template so I could try to follow this format.

For generating, it would need a way of incrementing a serial number, and then checking if that one exists and skipping if it does.

I am happy to give this a go myself, but from reading the documentation, it sounds like this discussion forum should be the first place to start, for ideas, tips and general assistance. Also some recommendations if this can be a plugin or if you would like it as a pull request.

Thank you for your time,
Harry.

**Comment by SergeoLacruz** [comment]
Hi,
Are you talking about internal part numbers? For IPN we use a strict format based on part class, housing and a number that is just counted upwards. We use an easy IPN generator that just looks for the next free number. The pattern is not configurable because we never change this. So it is specific for our use case. 
Michael

**Reply by readmodifywrite** [reply]
What easy IPN generator do you have?

**Reply by SergeoLacruz** [reply]
We use a simple python program that accesses the InvenTree database using the python library and prints out the next free number in the selected category.

**Reply by JacobEFO** [reply]
I suppose this is a proprietary tool, not one you could possibly share with the rest?

**Reply by SergeoLacruz** [reply]
Yes, you are right. Anyhow thee is not much to disclose. Just pick the next free numer on a certtain cat.

**Reply by JacobEFO** [reply]
It makes sense. Shouldn't be too hard once one has the hang of the plugins.

**Comment by Century1026** [comment]
I found [one](https://github.com/superbeppe98/ipn-generator)

**Comment by eeintech** [comment]
See https://github.com/inventree/InvenTree/discussions/6811
